modules/saml/lib/IdP/SAML2.php

```diff
@@ -986,9 +986,9 @@ private static function buildAssertion(
 
         if ($nameIdFormat === null || !isset($state['saml:NameID'][$nameIdFormat])) {
             // either not set in request, or not set to a format we supply. Fall back to old generation method
-            $nameIdFormat = $spMetadata->getString('NameIDFormat', null);
+            $nameIdFormat = array_values($spMetadata->getArrayizeString('NameIDFormat', null))[0];
             if ($nameIdFormat === null) {
-                $nameIdFormat = $idpMetadata->getString('NameIDFormat', \SAML2\Constants::NAMEID_TRANSIENT);
+                $nameIdFormat = array_values($idpMetadata->getArrayizeString('NameIDFormat', \SAML2\Constants::NAMEID_TRANSIENT))[0];
```

Comment:
If I may add a nitpick here:  I think array_shift() would be easier to read than array_values()[0] here

Reply:
> If I may add a nitpick here: I think array_shift() would be easier to read than array_values()[0] here

I've tended to avoid array_shift() recently because it sometimes has unexpected consequences that can be hard to debug when you use it on an lvalue. However, the construct with array_values() doesn't work with older PHP, so I guess here it probably makes more sense to do array_shift().

Reply:
> I think we should make it very clear in the documentation that this config option can now be an array, and that the first format specified in the array would be the one used by default.

I've added some documentation that hopefully does this